src/main/java/net/minestom/server/network/packet/client/handshake/HandshakePacket.java

```diff
@@ -92,8 +100,24 @@ public void process(@NotNull PlayerConnection connection) {
                         final String valueString = value.getAsString();
                         final String signatureString = signature == null ? null : signature.getAsString();
 
+                        if (BungeeCordProxy.isBungeeGuardEnabled() && nameString.equals("bungeeguard-token")) {
+                            if (foundBungeeGuardToken || !BungeeCordProxy.isValidBungeeGuardToken(valueString)) {
+                                socketConnection.sendPacket(new LoginDisconnectPacket(BungeeCordProxy.INVALID_TOKEN));
+                                socketConnection.disconnect();
+                                return;
+                            }
+
+                            foundBungeeGuardToken = true;
+                        }
+
                         properties.add(new GameProfile.Property(nameString, valueString, signatureString));
                     }
+
+                    if (BungeeCordProxy.isBungeeGuardEnabled() && !foundBungeeGuardToken) {
+                        socketConnection.sendPacket(new LoginDisconnectPacket(BungeeCordProxy.INVALID_TOKEN));
+                        socketConnection.disconnect();
```

Comment:
Should probably extract the disconnection into its own method as its re-used code 3 times

Reply:
what's the proper way to log something like this?

Reply:
I'm definitely not the right person to say what the *proper* way is, however most other classes seem to use     
`public final static Logger LOGGER = LoggerFactory.getLogger(YourClass.class);`
And some just use `MinecraftServer.LOGGER`
It's a bit inconsistent, but id personally probably go with the former rather than the latter.